Talk about key rotation downtime and "rotationPolicy: Always" (#521)
cert-manager/website: content/en/docs/usage/certificate.md

the [API reference documentation](../../reference/api-docs/#cert-manager.io/v1alpha2.CertificateSpec).

## X.509 key usages and extended key usages {#key-usages}

cert-manager supports requesting certificates that have a number of [custom key
usages](https://tools.ietf.org/html/rfc5280#section-4.2.1.3) and [extended key
usages](https://tools.ietf.org/html/rfc5280#section-4.2.1.12). Although
cert-manager will attempt to honor this request, some issuers will remove, add
defaults, or otherwise completely ignore the request.
The `CA` and `SelfSigned` `Issuer` will always return certificates matching the usages you have requested.

Unless any number of usages has been set, cert-manager will set the default
requested usages of `digital signature`, `key encipherment`, and `server auth`.
cert-manager will not attempt to request a new certificate if the current
certificate does not match the current key usage set.

An exhaustive list of supported key usages can be found in the [API reference
documentation](../../reference/api-docs/#cert-manager.io/v1alpha2.KeyUsage).

## Temporary Certificates while Issuing {#temporary-certificates-whilst-issuing}

> On old GKE versions (`1.10.7-gke.1` and below), when requesting certificates
> [using the ingress-shim](../ingress/) alongside the
> [`ingress-gce`](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress)
> ingress controller, `ingress-gce`
> [required](https://github.com/kubernetes/ingress-gce/pull/388) a temporary
> certificate is to be present while waiting for the issuance of a signed
> certificate. Note that this issue was
> [solved](https://github.com/jetstack/cert-manager/issues/606#issuecomment-424397233)
> in `1.10.7-gke.2`.

Reviewer: ```suggestion
On old GKE versions (`1.10.7-gke.1` and below), when requesting certificates
[using the ingress-shim](../ingress/) alongside the
[`ingress-gce`](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress)
ingress controller, `ingress-gce`
[required](https://github.com/kubernetes/ingress-gce/pull/388) a temporary
certificate must be present while waiting for the issuance of a signed
certificate. Note that this issue was
[solved](https://github.com/jetstack/cert-manager/issues/606#issuecomment-424397233)
in `1.10.7-gke.2`.
```

To work around this, you had to add the following annotation
to your Ingress objects:

```yaml
cert-manager.io/issue-temporary-certificate": "true"
```

That made sure that a temporary self-signed certificate was present in the
`Secret`. The self-signed certificate was replaced with the signed certificate
later on.

## Rotation of the private key {#rotation-private-key}

By default, the private key won't be rotated automatically. Using the setting
`rotationPolicy: Always`, the private key Secret associated with a Certificate
object can be configured to be rotated as soon as an action triggers the
reissuance of the Certificate object (see
[Actions that will trigger a rotation of the private key](#actions-triggering-private-key-rotation) below).

With `rotationPolicy: Always`, cert-manager waits until the Certificate
object is correctly signed before overwriting the `tls.key` file in the
Secret.

> With this setting, you can expect **no downtime** if your application can
> detect changes to the mounted `tls.crt` and `tls.key` using and reload them
> gracefully or automatically restart.

Reviewer: ```suggestion
With this setting, you can expect **no downtime** if your application can
detect changes to the mounted `tls.crt` and `tls.key` and reload them
gracefully or automatically restart.
```

If your application only loads the private key and signed certificate once
at start up, the new certificate won't immediately be served by your
application, and you will want to either manually restart your pod with
`kubectl rollout restart`, or automate the action by running
[wave](https://github.com/wave-k8s/wave). Wave is a Secret controller that
makes sure deployments get restarted whenever a mounted Secret changes.

{{% alert title="Re-use of private keys" color="primary" %}}

Some issuers, like the built-in [Venafi
issuer](/docs/configuration/venafi/), may disallow re-using private keys.
If this is the case, you must explicitly configure the `rotationPolicy:
Always` setting for each of your Certificate objects accordingly.

{{% /alert %}}

In the following example, the certificate has been set with
`rotationPolicy: Always`:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
spec:
  secretName: my-cert-tls
  privateKey:
    rotationPolicy: Always # 🔰 Here.
```

### Actions that will trigger a rotation of the private key {#actions-triggering-private-key-rotation}

> Setting the `rotationPolicy: Always` or change the Certificate's spec won't
> rotate the private key immediately. In order to get the private key secret
> rotated, the certificate objects must be reissued. A certificate object is
> reissued with either:

Reviewer: ```suggestion
Setting the `rotationPolicy: Always` or changing other fields in the Certificate spec won't
rotate the private key immediately. In order to rotate the private key, the certificate objects must be reissued. A certificate object is
reissued under the following circumstances:
```

Reviewer 2: Also:

> Changing the Certificate spec does not trigger the renewal process,

This statement in isolation is not true, and may cause confusion. The *majority* of fields *will* cause a re-issuance (e.g. changing dnsNames, commonName, usages etc). Changing the `rotationPolicy` (or any other fields that control how a renewal happens) *won't* cause a re-issuance immediately.

Author: In https://github.com/cert-manager/website/pull/521#discussion_r627557993, Richard proposed to remove the line:

> The certificate object is reissued when a change to the certificate object's spec is made.

I think I misunderstood Richard's suggestion back in May. I should have understood Richard's comment as:

> The certificate object is reissued when a change to one of the certificate object's fields `commonName`, `dnsNames`, `ipAddresses`, `uris`, `emailAddresses`, `subject`, `isCA`, `usages`, `duration` or `issuerRef` is made.

(as per [`RequestMatchesSpec`](https://github.com/jetstack/cert-manager/blob/8ae179b8f5aec708f8cb320a40ca8b495c9d889e/pkg/controller/certificates/util.go#L118))

Apologies for my misunderstanding.

I re-added the line about reissuing on spec change in ea1f25b.

{{% pageinfo color="warning" %}}

**❌** Deleting the Secret resource associated with a Certificate resource is
**not a recommended solution** for manually rotating the private key. The
recommended way to manually rotate the private key is to trigger the reissuance
of the Certificate resource with the following command (requires the [`kubectl
cert-manager`](/docs/usage/kubectl-plugin/#renew) plugin):

```sh
kubectl cert-manager renew cert-1
```

{{% /pageinfo %}}

### The `rotationPolicy` setting

The possible values for `rotationPolicy` are:

| Value                  | Description                                                   |
| ---------------------- | ------------------------------------------------------------- |
| `Never` (default)      | cert-manager reuses the existing private key on each issuance |
| `Always` (recommended) | cert-manager regenerates a new private key on each issuance   |

With `rotationPolicy: Never`, a private key is only generated if one does not
already exist in the target Secret resource (using the `tls.key` key). All
further issuances will re-use this private key. This is the default in order to
maintain compatibility with previous releases.

> With `rotationPolicy: Always`, a new private key will be generated each time an
> action triggers the reissuance of the certificate object (see
> [actions](#actions-triggering-private-key-rotation) above). Note that if the
> private key secret already exists when creating the certificate object, the
> existing private key will not be used, since the rotation mechanism also
> includes the initial issuance.

Reviewer: ```suggestion
With `rotationPolicy: Always`, a new private key will be generated each time an
action triggers the reissuance of the certificate object (see
[Actions that will trigger a rotation of the private key](#actions-triggering-private-key-rotation) above). Note that if the
private key secret already exists when creating the certificate object, the
existing private key will not be used, since the rotation mechanism also
includes the initial issuance.
```

{{% pageinfo color="info" %}}

> 👉 It is recommended you configure `rotationPolicy: Always` on your certificate
> resources. It is considered to be a good practice to rotate private keys when a
> certificate is renewed.

Reviewer: ```suggestion
👉 We recommend that you configure `rotationPolicy: Always` on your certificate
resources. It is considered to be a good practice to rotate private keys when a
certificate is renewed.
```

Can we link to a reference backing up this claim good practice?

Author: I cannot find any good reference to support this claim. By reading some discussions:

- https://community.letsencrypt.org/t/should-i-always-generate-a-new-csr-private-key/100985
- https://security.stackexchange.com/questions/27810/should-i-change-the-private-key-when-renewing-a-certificate

People seem to suggest that:
1. RSA/ECDSA private keys can't (mostly) be guessed, which means there is less of a need for renewing them (unlike passwords that are way shorter),
2. But at the same time, an RSA/ECDSA private keys may be leaked through some old backup, leading to a possible man-in-the-middle situation.

Looking around, I found that:

- certbot renews the private key while renewing the cert by default, but there is nothing explaining why;
- NIST's [Guidelines for the Selection, Configuration, and Use of Transport Layer Security (TLS) Implementations](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-52r2.pdf) does not indicate when to renew the private key; the only bit that talks about what to do with a server private key is this:
  > The security of the server’s private key is critical to the security of TLS. If the server’s private key is weak or can be obtained by a third party, the third party can masquerade as the server to
all clients.
- NIST's [Recommendation for Key Management](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf) has a nice table but no claim on when the private key should be renewweed:
  ![image](https://user-images.githubusercontent.com/2195781/127500092-d779df24-8c3c-445d-a092-4e119757ce83.png)

- f5-bigip suggests to renew the private key (https://support.f5.com/csp/article/K14620):

  > Note: When renewing an SSL certificate from a CA, F5 recommends that you generate a new certificate signing request (CSR) and private key. Although some CAs allow you to renew a certificate by using the existing CSR on file, this method is less secure as it retains the existing private key. To generate a new CSR, follow the procedure for Creating an SSL CSR and private key.

@SgtCoDFish Do you know a place where that would be written? Or should I rather remove this claim and say that "the cert-manager team suggests users renewing the private key along with the certificate"?

Reviewer 2: I think one big reason why you _would_ want to, is that if you were to renew a certificate using the same private key, and then after the initial certificates expiry learn that the private key had actually been leaked, the new certificate would be valid to use with the old private key, thus the 'rotation' didn't really provide any more protection and really just extended the lifetime of the existing certificate.

Reviewer 3: James is spot on. I don't have like an authoritative source to hand for why you'd want to rotate as much as possible, but certainly I think it's best practice. 

The longer a key is used, the more likely it is to be exposed. Keys should be renewed "as often as possible" for me; not only for security reasons but also operational reasons to show that it can be done, so that if it has to be done in an emergency the org has confidence that they can carry it out.

Author: I tried my best to formulate a paragraph that "justifies" our point of view. Please forgive me if any of this is untrue. 🔥

👉 We recommend that you configure `rotationPolicy: Always` on your Certificate resources. 

Rotating both the certificate and the private key simultaneously prevents the risk of issuing a certificate associated with an exposed private key.

Another benefit to renewing the private key regularly is to let you be confident that the private key rotation can be done in case of an emergency.

More generally, it is a good practice to be rotating the keys as often as possible, reducing the risk associated with compromised keys.

usages](https://tools.ietf.org/html/rfc5280#section-4.2.1.3) and [extended key
usages](https://tools.ietf.org/html/rfc5280#section-4.2.1.12). Although
cert-manager will attempt to honor this request, some issuers will remove, add
defaults, or otherwise completely ignore the request and is determined on an
issuer by issuer basis. The `CA` and `SelfSigned` `Issuer` will always return
certificates matching the usages you have requested.

Unless any number of usages has been set, cert-manager will set the default
requested usages of `digital signature`, `key encipherment`, and `server auth`.
cert-manager will not attempt to request a new certificate if the current
certificate does not match the current key usage set.

An exhaustive list of supported key usages can be found in the [API reference
documentation](../../reference/api-docs/#cert-manager.io/v1alpha2.KeyUsage).

## Temporary Certificates while Issuing

When requesting certificates [using the ingress-shim](../ingress/), the
component `ingress-gce`, if used, requires that a temporary certificate is
present while waiting for the issuance of a signed certificate when serving. To

> facilitate this, if the following annotation:

Author: I found two places where this "trick" is mentioned:

- the actual PR where this annotation was added: https://github.com/jetstack/cert-manager/pull/2089,
- In the [release notes for 0.11](https://github.com/jetstack/cert-manager/blob/dffbf391dbb0fc6c1cfea62e561a9c6f54362ab0/design/release-notes/release-0.11/draft-release-notes.md#no-longer-generating-temporary-certificates-by-default):

  > **New CertificateRequest resource type**
  >
  > The activation of `CertificateRequest` controllers are no longer behind a feature flag and are now instead enabled by default.
  >
  > This means that when requesting certificates using the `Certificate` resource the `CertificateRequest` resource will be used as the default and only way to honour the request.
  >
  > The addition of this resource introduces the ability for much greater extension points to cert-manager, notably out-of-tree issuers, istio integrations, and experimental tooling such as a CSI driver. You can read more about the motivation and design of this resource in the [enhancement document](https://github.com/jetstack/cert-manager/blob/master/design/20190708.certificate-request-crd.md).
  >
  > This change should cause no disruption to how end users interact with cert-manager, with the exception of debugging now requiring this resource to be inspected also.

- But by far the best explanation has been written by James Munnelly in https://github.com/kubernetes/ingress-gce/pull/388:

  > This switches ingress-gce back to the previous behaviour of 'soft failing' if the specified TLS secret exists. Without this change, we've had many users complain that auto-TLS using cert-manager (and also kube-lego) is broken.

I'll go ahead and add a link to this ingress-gce PR.

> facilitate this, if the following annotation:

Author: I now use the past tense since this seems to be a solved issue:

> On old GKE versions (1.10.7-gke.1 and below), when requesting certificates [using the ingress-shim](../ingress/) alongside the [`ingress-gce`](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress) ingress controller, `ingress-gce` [required](https://github.com/kubernetes/ingress-gce/pull/388) a temporary certificate is to be present while waiting for the issuance of a signed certificate. Note that this issue was solved [in 1.10.7-gke.2](https://github.com/jetstack/cert-manager/issues/606#issuecomment-424397233).
> 
> To work around this, you had to add the following annotation to your Ingress objects:
>
> ```yaml
> cert-manager.io/issue-temporary-certificate": "true"
> ```
>
> That made sure that a temporary self-signed certificate is present in the `Secret`. The self-signed certificate is replaced with the properly signed certificate later on.

@munnerz I am correct in saying that this annotation is not needed anymore?

> facilitate this, if the following annotation:

Author: I will resolve this conversation in order to move this PR forward. Please re-open it if you think I missed something.

```yaml
cert-manager.io/issue-temporary-certificate": "true"
```

is present on the certificate, a self-signed temporary certificate will be
present on the `Secret` until it is overwritten once the signed certificate has
been issued.

## Rotation of the private key {#rotation-private-key}

By default, the private key won't be rotated automatically. Using the setting
`rotationPolicy: Always`, the private key secret associated with a certificate
object can be configured to be rotated as soon as an action triggers the
reissuance of the certificate object (see
[actions](#actions-triggering-private-key-rotation) below).

With this setting, cert-manager waits until the certificate object is correctly
signed before overwriting the private key secret, which means you can expect

> **no downtime** with regards to the private key secret.

Author: Good point, I totally forgot about the fact that "no downtime" also implies that any application using the secret immediately uses the new certificate...

I added your suggestion in 62a8bf15108e7f21c0dd675f268c5575cd70b738. Thank you!!

> **no downtime** with regards to the private key secret.

Author: I will resolve this conversation in order to move this PR forward. Please re-open it if you think I missed something.

Some Issuer types may disallow re-using private keys. If this is the case, you
must explicitly configure the `rotationPolicy: Always` setting for each of your

> certificates objects accordingly.

Author: I have been looking around (searched "reuse private key" on GitHub Search) but I wasn't able to find anything.

I found out about [ReusePrivateKey](https://github.com/jetstack/cert-manager/blob/2aa625edcb1a179600d4c557216c7bfeb44ae269/test/e2e/framework/helper/featureset/featureset.go#L106):

```go
	// ReusePrivateKey denotes whether the target issuer is able to sign multiple
	// certificates for the same private key. This is useful for some issuers
	// that have trouble being configured to support this feature.
	ReusePrivateKeyFeature Feature = "ReusePrivateKey"
```

After some git wrangling, I found the commit https://github.com/jetstack/cert-manager/commit/a96f93e0c9bc9b38fa1804f557707dbe9f3b3418 that introduced the `ReusePrivateKey` flag. The commit message points towards the Venafi issuer, I'll add the venafi issuer as an example.

> certificates objects accordingly.

Author: Done in 9371683!

In the following example, the certificate has been set with `rotationPolicy: Always`:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
spec:
  secretName: my-cert-tls
  privateKey:
    rotationPolicy: Always # 🔰 Here.
```

### Actions that will trigger a rotation of the private key {#actions-triggering-private-key-rotation}

Setting the `rotationPolicy: Always` won't do anything by itself. In order to
get the private key secret rotated, the certificate objects must be reissued. A
certificate object is reissued with either:

- when the X.509 certificate is nearing expiry,

> - when a change to the certificate object's spec is made,

Author: Good point, I made it clearer that a change to the certificate does not trigger a renew of the private key in ee43f0c.

- when a reissuance is manually triggered with the following:
  ```sh
  kubectl cert-manager renew cert-1
  ```
  Note that the above command requires the [kubectl
  cert-manager](/docs/usage/kubectl-plugin/#renew) plugin.

{{% pageinfo color="warning" %}}

> **🛑** Deleting the Secret resource associated a the Certificate resource is
> **not a recommended solution** for manually rotating the private key. The
> recommended way to manually rotate the private key is to trigger a reissuance of
> the Certificate resource with the following command (requires the [kubectl
> cert-manager](/docs/usage/kubectl-plugin/#renew) plugin):

Reviewer: ```suggestion
**🛑** Deleting the Secret resource associated a Certificate resource is
**not a recommended solution** for manually rotating the private key. The
recommended way to manually rotate the private key is to trigger a reissuance of
the Certificate resource with the following command (requires the [kubectl
cert-manager](/docs/usage/kubectl-plugin/#renew) plugin):
```

```sh
kubectl cert-manager renew cert-1
```

{{% /pageinfo %}}

### The `rotationPolicy` setting

The possible values for `rotationPolicy` are:

| Value                  | Description                                                   |
| ---------------------- | ------------------------------------------------------------- |
| `Never` (default)      | cert-manager reuses the existing private key on each issuance |
| `Always` (recommended) | cert-manager regenerates a new private key on each issuance   |

> With `rotationPolicy: Never`, a private key is only generated if one does not
> already exist in the target Secret resource (using the `tls.key` key). All
> further issuance's will re-use this private key. This is the default in order to
> maintain compatibility with previous releases.

Reviewer: ```suggestion
With `rotationPolicy: Never`, a private key is only generated if one does not
already exist in the target Secret resource (using the `tls.key` key). All
further issuances will re-use this private key. This is the default in order to
maintain compatibility with previous releases.
```

A full list of the fields supported on the Certificate resource can be found in
the [API reference documentation](../../reference/api-docs/#cert-manager.io/v1alpha2.CertificateSpec).

## X.509 key usages and extended key usages {#key-usages}

cert-manager supports requesting certificates that have a number of [custom key
usages](https://tools.ietf.org/html/rfc5280#section-4.2.1.3) and [extended key
usages](https://tools.ietf.org/html/rfc5280#section-4.2.1.12). Although
cert-manager will attempt to honor this request, some issuers will remove, add
defaults, or otherwise completely ignore the request.
The `CA` and `SelfSigned` `Issuer` will always return certificates matching the usages you have requested.

Unless any number of usages has been set, cert-manager will set the default
requested usages of `digital signature`, `key encipherment`, and `server auth`.
cert-manager will not attempt to request a new certificate if the current
certificate does not match the current key usage set.

An exhaustive list of supported key usages can be found in the [API reference
documentation](../../reference/api-docs/#cert-manager.io/v1alpha2.KeyUsage).

> ## Temporary Certificates while Issuing

Author: I just realized that this change is a breaking change for anyone who use the anchor ID, e.g.

> https://cert-manager.io/docs/usage/certificate/#temporary-certificates-whilst-issuing

Should I use instead:

```md
## Temporary Certificates while Issuing {#temporary-certificates-whilst-issuing}
```

to avoid breaking the anchor ID?
